### PR TITLE
feat: add apisix-openresty deb support

### PR DESCRIPTION
**Pull request by EnableAsync**
Issue: #62

### CODE REVIEW

**Comment by EnableAsync** [comment]
Ping @spacewander @tzssangglass @imjoey for review at your free time. Thanks.

File: .github/workflows/package-apisix-openresty-rpm-el7.yml [hunk 1]
```diff
@@ -21,8 +21,6 @@ jobs:
       - name: install dependencies
         run: |
           sudo apt-get install -y make ruby ruby-dev rubygems build-essential
-          sudo gem install --no-document fpm
-          sudo apt-get install -y rpm
```
Comment (imjoey) [inline .github/workflows/package-apisix-openresty-rpm-el7.yml:25]: @EnableAsync any reason to remove these two lines?

Comment (EnableAsync) [reply .github/workflows/package-apisix-openresty-rpm-el7.yml:25]: Now that fpm is installed in docker, there is no need to install it in github action.

File: Makefile [hunk 2]
```diff
@@ -46,15 +47,20 @@ ifneq ($(buildx), True)
 define build
 	docker build -t apache/$(1)-$(3):$(version) \
 		--build-arg checkout_v=$(checkout) \
+		--build-arg version=$(version) \
+		--build-arg debug=$(debug) \
 		--build-arg IMAGE_BASE=$(image_base) \
 		--build-arg IMAGE_TAG=$(image_tag) \
 		--build-arg CODE_PATH=$(4) \
+		--build-arg DEBUG=$(debug) \
```
Comment (imjoey) [inline Makefile:55]: @EnableAsync two arguments for opt-in the `debug` variable.

File: dockerfiles/Dockerfile.apisix-openresty.rpm [hunk 3]
```diff
@@ -3,19 +3,21 @@ ARG IMAGE_TAG="7"
...
 
-COPY build-apisix-openresty-centos7.sh /tmp/build-apisix-openresty-centos7.sh
+COPY ./utils/build-common.sh /tmp/build-common.sh
 COPY build-apisix-openresty.sh /tmp/build-apisix-openresty.sh
 COPY ./utils/determine-dist.sh /tmp/determine-dist.sh
 
 WORKDIR /tmp
 
 ARG version
+ARG debug
```
Comment (imjoey) [inline dockerfiles/Dockerfile.apisix-openresty.rpm:13]: For arguments defined in dockerfile, it's common to use in upper characters.

**Review by imjoey — CHANGES_REQUESTED** [review]
Hi @EnableAsync, thank you very much for this PR. Generally ok for me. While I noticed that this PR is going to make two kinds of changes:

1. Add support for generating the deb format of apisix-openresty package
2. Add support for generating the debug version of apisix-openresty packages

It would be more preferable that one PR only does one thing. So could you please only contain the 1st change? Then we can put the 2nd change in other PRs.

Thanks.

File: Makefile (hunk 2 above)
Comment (EnableAsync) [reply Makefile]: Thanks for your kind reminder, it has been updated.

File: dockerfiles/Dockerfile.apisix-openresty.rpm (hunk 3 above)
Comment (EnableAsync) [reply dockerfiles/Dockerfile.apisix-openresty.rpm]: Much appreciated your kind reminder. XD

File: .github/workflows/package-apisix-openresty-rpm-el7.yml (hunk 1 above)
Comment (imjoey) [reply .github/workflows/package-apisix-openresty-rpm-el7.yml:25]: @EnableAsync Got it. Thanks.

File: Makefile [hunk 4]
```diff
@@ -46,6 +46,7 @@ ifneq ($(buildx), True)
 define build
 	docker build -t apache/$(1)-$(3):$(version) \
 		--build-arg checkout_v=$(checkout) \
+		--build-arg version=$(version) \
```
Comment (imjoey) [inline Makefile:49]: @EnableAsync Please use the upper case `VERSION` here. Thanks.

File: Makefile [hunk 5]
```diff
@@ -55,6 +56,7 @@ else
 define build
 	docker buildx build -t apache/$(1)-$(3):$(version) \
 		--build-arg checkout_v=$(checkout) \
+		--build-arg version=$(version) \
```
Comment (imjoey) [inline Makefile:59]: Ditto. Thanks.

File: dockerfiles/Dockerfile.apisix-openresty.deb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+FROM ${IMAGE_BASE}:${IMAGE_TAG}
+
+COPY ./utils/build-common.sh /tmp/build-common.sh
+COPY build-apisix-openresty.sh /tmp/build-apisix-openresty.sh
+COPY ./utils/determine-dist.sh /tmp/determine-dist.sh
+
+WORKDIR /tmp
+
+ARG version
```
Comment (imjoey) [inline dockerfiles/Dockerfile.apisix-openresty.deb:12]: Ditto, `VERSION` here.

File: dockerfiles/Dockerfile.apisix-openresty.deb [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+ARG IMAGE_BASE
+ARG IMAGE_TAG
+
+ENV IMAGE_BASE=${IMAGE_BASE}
+ENV IMAGE_TAG=${IMAGE_TAG}
+
+RUN version=${version} \
```
Comment (imjoey) [inline dockerfiles/Dockerfile.apisix-openresty.deb:19]: Ditto, it should be `VERSION=${VERSION}`. Thanks.

File: dockerfiles/Dockerfile.apisix-openresty.rpm [hunk 8]
```diff
@@ -16,6 +16,7 @@ ARG IMAGE_TAG
 ENV IMAGE_BASE=${IMAGE_BASE}
 ENV IMAGE_TAG=${IMAGE_TAG}
 
-RUN version=${version} ./build-apisix-openresty-centos7.sh \
+RUN version=${version} \
```
Comment (imjoey) [inline dockerfiles/Dockerfile.apisix-openresty.rpm:19]: Ditto here.

File: Makefile (hunk 4 above)
Comment (EnableAsync) [reply Makefile]: Thanks for your kind reminder, updated.

File: Makefile (hunk 5 above)
Comment (EnableAsync) [reply Makefile]: Updated.

File: dockerfiles/Dockerfile.apisix-openresty.deb (hunk 6 above)
Comment (EnableAsync) [reply dockerfiles/Dockerfile.apisix-openresty.deb]: Updated.

File: dockerfiles/Dockerfile.apisix-openresty.deb (hunk 7 above)
Comment (EnableAsync) [reply dockerfiles/Dockerfile.apisix-openresty.deb]: Changed to use `ENV` to set environment variables.

File: dockerfiles/Dockerfile.apisix-openresty.rpm (hunk 8 above)
Comment (EnableAsync) [reply dockerfiles/Dockerfile.apisix-openresty.rpm]: Changed to use `ENV` to set environment variables.

**Comment by EnableAsync** [comment]
Regarding the reason for the CI failure, I think we should use `yum install epel-release` to install epel-release, which was verified here[1].
Looking forward to your suggestions.

[1] https://github.com/EnableAsync/apisix-build-tools/runs/3583747375?check_suite_focus=true

**Comment by imjoey** [comment]
> Regarding the reason for the CI failure, it seems that the download link for `epel-release` is no longer available, and I think we should use `yum install epel-release` to install epel-release, which was verified here[1].
> Looking forward to your suggestions.
> 
> [1] https://github.com/EnableAsync/apisix-build-tools/runs/3583747375?check_suite_focus=true

@EnableAsync you're right. That would eliminate this CI failure. Would you please submit another PR for this now? We would better fix the CI failures before reviewing/merging this PR. What do you think? Thanks.

**Comment by EnableAsync** [comment]
> > Regarding the reason for the CI failure, it seems that the download link for `epel-release` is no longer available, and I think we should use `yum install epel-release` to install epel-release, which was verified here[1].
> > Looking forward to your suggestions.
> > [1] https://github.com/EnableAsync/apisix-build-tools/runs/3583747375?check_suite_focus=true
> 
> @EnableAsync you're right. That would eliminate this CI failure. Would you please submit another PR for this now? We would better fix the CI failures before merging this PR. What do you think? Thanks.

@imjoey Hi, I have submitted the pr(https://github.com/api7/apisix-build-tools/pull/93), please take a look in your free time. Thanks.

**Comment by imjoey** [comment]
> > > Regarding the reason for the CI failure, it seems that the download link for `epel-release` is no longer available, and I think we should use `yum install epel-release` to install epel-release, which was verified here[1].
> > > Looking forward to your suggestions.
> > > [1] https://github.com/EnableAsync/apisix-build-tools/runs/3583747375?check_suite_focus=true
> > 
> > 
> > @EnableAsync you're right. That would eliminate this CI failure. Would you please submit another PR for this now? We would better fix the CI failures before merging this PR. What do you think? Thanks.
> 
> @imjoey Hi, I have submitted the pr(#93), please take a look in your free time. Thanks.

@EnableAsync #93 has been merged. Please rebase codes here. Thanks. 😄

**Review by imjoey — APPROVED** [review]
/lgtm

**Comment by imjoey** [comment]
@EnableAsync thank you very much for this. 😄

**Comment by imjoey** [comment]
@EnableAsync Hi, the readme seems not updated yet. Would you please submit another PR to add relevant instructions? Thanks.

**Comment by EnableAsync** [comment]
> @EnableAsync Hi, the readme seems not updated yet. Would you please submit another PR to add relevant instructions? Thanks.

@imjoey Hi, I have submitted the pr(#94), please take a look in your free time. Thanks.

**Pull request merged**